File: 10-week-8--searching/02-day-5--build-a-binary-tree---balancing/javascript/tests/validation_methods.js
function bstToArray(root) {
  if (root === null) {
    return [];
  }

  const stack = [root];
  const sorted = [];

  while (stack.length) {
    const node = stack[stack.length - 1];

    if (node.left !== null) {
      stack.push(node.left);
      node.left = null;
      continue;
    }

    sorted.push(stack.pop().value);

    if (node.right !== null) {
      stack.push(node.right);
    }
  }

  return sorted;
}

function isBalanced(root) {
  if (root === null) {
    return true;
  }

  const leftHeight = branchHeight(root.left);
  const rightHeight = branchHeight(root.right);

  return Math.abs(leftHeight - rightHeight) < 2;
}

function branchHeight(root) {
  let queue = root === null ? [] : [root];
  let count = 0;

  while(queue.length) {
    ++count;

    queue = queue.reduce((accum, node) => {
      if (node.left) {
        accum.push(node.left);
      }

      if (node.right) {
        accum.push(node.right);
      }

      return accum;
    }, []);
  }

  return count;
}

module.exports = {
  bstToArray,
  isBalanced
};

File: 10-week-8--searching/02-day-5--build-a-binary-tree---balancing/ruby/spec/validation_methods.rb
class Validator
  def self.bst_to_array(root)
    if (root == nil)
      return []
    end
  
    stack = [root]
    sorted = []
  
    while stack.length > 0
      node = stack.last
  
      if (node.left != nil)
        stack.push(node.left)
        node.left = nil
        next
      end
  
      sorted.push(stack.pop().value)
  
      if (node.right != nil) 
        stack.push(node.right)
      end
    end
  
    sorted
  end

  def self.balanced?(root)
    return true if root.nil?
    
    left_height = branch_height(root.left)
    right_height = branch_height(root.right)

    (left_height - right_height).abs < 2
  end

  def self.branch_height(root)
    queue = root.nil? ? [] : [root]
    count = 0

    until queue.empty?
      count += 1

      queue = queue.reduce([]) do |accum, node|
        accum << node.left if node.left
        accum << node.right if node.right
        accum
      end
    end

    count
  end
end


